File: WallBreakers/Arrays/SortArrayByParity.java
class Solution {
    public int[] sortArrayByParity(int[] A) {
        int h = 0, t = A.length-1;
        int temp;
        while (h < t) {
            if (A[h] % 2 != 0 && A[t] % 2 == 0) { // h is odd, t is even
                // swap
                temp = A[h];
                A[h] = A[t];
                A[t] = temp;
            }
            
            if (A[h] % 2 == 0) h++;
            if (A[t] % 2 != 0) t--;
        }
        return A;
    }
}

File: WallBreakers/Arrays/SortArrayByParity.py
class Solution:
    def sortArrayByParity(self, A: List[int]) -> List[int]:
        h, t = 0, len(A) - 1
        while h < t:
            if A[h] % 2 != 0 and A[t] % 2 == 0:
                A[h], A[t] = A[t], A[h]

            if A[h] % 2 == 0:
                h += 1
            if A[t] % 2 != 0:
                t -= 1

        return A
